### ISSUE TITLE
StyleGAN→BigGAN: import the StyleGAN large 8x512 FC _z_ → _w_ embedding trick

### ISSUE THREAD

**Issue by gwern** [tensorfork/tensorfork#26]
It would be nice to make BigGAN more like StyleGAN in a few ways, primarily its stability. StyleGAN also seems to have a better latent space with more disentangled and linearized factors, which is great for editing but may also play a role in the stability. 

As far back as IllustrationGAN, people have noted that feeding the _z_ embedding (which is typically just a bunch of random normals, like 128 N(0,1)s) straight into the CNN doesn't work as well as feeding _z_ into 1 or 2 FC layers first. Me & FeepingCreature noticed this for WGAN as well. BigGAN experimented with a variety of other starting distributions, but kept N(0,1) for simplicity. StyleGAN took the breathtaking step of plopping in no less than *8* FC layers to transform _z_ into... something, before passing the final resulting _w_ into the rest of StyleGAN. (It's also worth noting that StackGAN's implementation appears to pass its text embeddings through at least 1 FC layer as part of its "noise augmentation", which may be part of why its text→image works and [ours doesn't](https://github.com/tensorfork/tensorfork/issues/10).) The theory is that the 'true' latent space of the data distribution is extremely nonlinear and complex and non-normal and that starting with N(0,1)s is extremely unhelpful to the convolution layers trying to create something realistic, but that passing _z_ through a deep stack of FC layers lets it be massaged into something that encodes in an easy-to-understand-way everything the rest of StyleGAN needs to know, and reduces the need for things like global attention layers to enforce consistency.

Perhaps this would be useful for BigGAN? We can simply paste in an 8x512 block after _z_ and see how it goes. As tweaks go, this one should be *very* easy to do and potentially quite helpful, and so is relatively high priority. If it works well, we can consider whether self-attention layers would work even better. (The better this works, the more evidence it provides for the idea of an all-attention GAN.)

**Comment by aydao** [comment]
It is mostly straightforward to add a mapping (sub)network to the generator of BigGAN. A first attempt is [here](https://github.com/aydao/compare_gan/blob/aymapping/compare_gan/architectures/resnet_biggan_mapping.py#L244). The major change is to inject eight fully-connected sequential layers (using leaky ReLu, smaller learning rates) between the input of z to G and the point at which z is broadcast to later layers. In essence, `z_in -> mapping subnetwork -> 'z' (a.k.a 'w' in StyleGAN) -> normal BigGAN G`. Additionally, a choice must be made to either disable hierarchical z or rework how chunks of z are computed.

Open questions include:

- What dimension of the latent space (z_dim) is appropriate to use?
- What is the appropriate learning rate multiplier? Karras et al. use 0.01 relative to the rest of the network, which is likely selected empirically.
- How substantial are the effects of moving from hierarchical z to the simpler scheme of just passing along the entire z? BigGAN chunks z and passes each chunk to a respective layer; StyleGAN and BigGAN-deep just pass all of the latent to the intermediate synthesis layers.
- StyleGAN uses style-mixing (along with learned affine transformations, AdaIn / demodulation...). How much of those will be necessary for BigGAN with a network?
- StyleGAN tracks the average of w (dlatent_avg in StyleGAN) for its truncation trick. Is a similar process required / useful for BigGAN with a mapping network? 
- StyleGAN applies normalization to the latents prior to the mapping network. Is normalization necessary for BigGAN's mapping network? I have tested StyleGAN without normalization on numerous datasets including FFHQ, and it is not necessary to achieve models capable of high quality.

**Comment by gwern** [comment]
Update: aydao has experimented with doing 120→512✕8→120 and 120→120✕8→120.

The first two experiments wound up diverging despite using the Karras 0.01 multiplier, but this appears to have been due to forgetting to enable spectral normalization (without which all of BigGAN is unstable, so the _w_ stack turns out to be no exception, unsurprisingly).

Turning spec-norm on for 120→120✕8→120 appears to be working well as of [40k iterations](https://cdn.discordapp.com/attachments/693736350561861633/709486213655953518/individualImage.png); interestingly, the EMA samples, which usually give a noticeable upgrade in quality by 40k, are still garbage although the losses and spec-norm graphs all look well-behaved and the samples look quite nice, suggesting that the model is continuing to learn and evolve rapidly past the point where most BigGANs have settled down a bit.

Next thing to try is returning to the 120→512✕8→120 _w_ architecture, since BigGAN should benefit from the more powerful embedding, especially as we train on ImageNet, Danbooru, and combinations of datasets ('chaos').

---

Bumping to high priority because it's proven easy to get working, so we should get this in as part of our defaults ASAP to benefit from it, especially for [chaos runs](https://github.com/tensorfork/tensorfork/issues/28).

**Comment by aydao** [comment]
There are several positive developments and observations to detail. For now, it's worth recording remarks to the earlier open questions:

- What dimension of the latent space (z_dim) is appropriate to use? **I've tried both z_dim and 512.** The former seems more natural as the learned mapping is never overparameterized, however having 8x512 layers appears to grant G more representative capacity, with that run leading to G decisive winning over D.
- What is the appropriate learning rate multiplier? **I've tried both 0.01 and 1.0,** and the latter has no noticeable negative impact on performance.
- How substantial are the effects of moving from hierarchical z to the simpler scheme of just passing along the entire z? **For now, this architecture just uses hierarchical z.** This question is worth revisiting if/when this idea is applied to BigGAN-deep.
- How much of (style mixing, normalization, demodulation) will be necessary for BigGAN with a network? **None appear *necessary*.** The architecture appears to be getting many StyleGAN-type benefits already, so if anything it may be good to explore ways this architecture could be more like BigGAN.
- Is (tracking w_avg) required / useful for BigGAN with a mapping network? **Unknown as yet.** The EMA samples are almost invariably blobby (completely opposite vanilla BigGAN), though some reflect stable patterns that appear in the non-EMA instances.

**Comment by gwern** [comment]
How I would summarize the past few days: _w_ works in BigGAN and yields stable training without any LR tricks; _w_, however, does seem to come at a cost to quality because it moves too fast for the rest of the G to catch up, as can be shown by freezing _w_, at which point the G samples rapidly improve in sharpness. EMA remains blurry, however.

Why? EMA is implemented properly and includes _w_ as expected; however, StyleGAN uses by default a much faster decay than stock BigGAN (attempting to scale it with bigger BigGAN-like minibatches), and _w_ BigGAN needs to likewise use <<0.99, and indeed, using EMA decays like 0.7 work much better than the stock BigGAN EMA decay hyperparameter of 0.999. This seems to be normal and expected.

So we have established that _w_ doesn't kill BigGAN, and how to fix EMA for _w_. the remaining mystery is how to deal with the training sample degradation. Is _w_ inherently harmful? Or do we need to just use the LR multiplier? If StyleGAN, with such restricted domains and a const input, needs a LR mulipltier of 0.01, maybe BigGAN needs 0.001. We showed that BigGAN works with 1 in the sense that it doesn't destabilize or diverge, but the samples kept changing rapidly and never sharpened up within the first ~150k iterations.

So my next suggestion is after setting EMA decay to something like 0.8-0.9, we try going back to the LR multiplier in a long-trained model (100k+) but with 0.001.

A very deep FC stack can create extremely nonlinear changes through itself with relatively small adjustments, so maybe that's what's going on, is all. We just haven't been aggressive enough in reducing the LR multiplier to fit the extremely challenging task of BigGAN.

**Comment by gwern** [comment]
We're not seeing any clear improvement in EMA over training. Some are better, some are different are worse. Around 404k iterations, it's doing better in terms of global structure, but that's still extremely slower than regular BigGAN.

I wonder what we're missing. It seems unlikely BigGAN could be disentangling just as well as StyleGAN all on its own, could it?
Maybe the FC stack is too deep. The self-attention should be covering part of its duties already. Maybe we should be trying something simpler like 2 layers instead of 8
Right now, we haven't seen any particularly obvious gains from any version of _w_, have we?
We admittedly are not calculating any FIDs or path lengths like we should be, but the gains are not exactly jumping out in the samples.
We could just anneal _w_ as a way of LR scheduling, but it's not clear right now that we get any benefit at all from _w_, is it?

Shawn is a little obsessed with the Laplacian pyramid of StyleGAN vs the GAN/BigGAN implicit upscaling approach. I wonder if _w_ is not working for us because BigGAN implicitly is already doing all that disentangling spread throughout the G?
That is, StyleGAN has to have the image globally specified as soon as _w_ starts going into the AdaIN blocks, because the immediate output of the very first block is going to be added to the final output along with all of the outputs of successive blocks. StyleGAN is 'committed' early on, there's no 'time [layers] to think' and do the equivalent of [remolding the original Gaussian](https://arxiv.org/pdf/1812.04948.pdf#page=6 "Figure 6. Illustrative example with two factors of variation (image features, e.g., masculinity and hair length)"). BigGAN has the self-attention, but it can continue tweaking everything until the very last block when it finally commits to specific pixels, right?

1. falsifiable prediction: if we converted BigGAN to Laplacian Pyramid style image generation with only _z_, it'd yield disappointing results until _w_ was added
2. another prediction: if _w_ is doing work previously spread out through BigGAN, we should see the post-_w_ layers be 'simpler' in some sense thanks to it. Smaller average specnorms? More linear matrices?
3. ancillary prediction: if _w_ is 'doing more work upfront', then that makes changes cascade, and we should be lowering the LR of the entire BigGAN, as G can now pivot much faster with changes inside w and minimal changes elsewhere, and it needs to be harder to make changes in the rest of BigGAN as well (if you think of _w_ as a 'seed' with the rest of BigGAN learning 'modules' like 'green eye', small changes in _w_ can greatly change the balance of emitted images even while holding the rest of BigGAN entirely constant. but the gradients right now will change both _w_ and the rest of BigGAN)

So current suggestions: try runs with fewer _w_ layers like 2; compare the specnorms of the deep _w_ runs with comparable non-_w_ runs; try lowering the global LR substantially like 10x; try adding some FC layers to D before its projection/classifier heads to give it a _w_ to match G (D's channel count has already been boosted to 120 to keep up with _w_-G but since the G tends to have a loss at ~0.5 and D is >1, it may still not be balanced.

**Comment by gwern** [comment]
It seems with #34 we may have accidentally solved the _w_ instability issue. You just need a more useful mix of latent _z_ variables to break the symmetry and reflect sparsity/discreteness.

**Comment by gwern** [comment]
Another way to fix _w_ is to initialize it better in the sense of residual/skip connections. [ReZero](https://arxiv.org/pdf/2003.04887.pdf) is an example of how deep FC stacks (well beyond 8 layers, up to 1k+) can be trained stably and easily by adding in residual connections so the input doesn't get completely scrambled by the extremely nonlinear random initialization.
